src/dlsproto/client/request/internal/GetRange.d

```diff
@@ -744,41 +792,56 @@ private scope class GetRangeHandler
                     Message(MessageType_v2.Stopped),
-                    Message(MessageType_v2.Finished));
+                    Message(MessageType_v2.Finished),
+                    Signal(this.outer.TimeoutSignal));
 
-                switch (msg.type)
+                with (event.Active) switch (event.active)
                 {
-                    case MessageType_v2.Records:
-                        Const!(void)[] received_record_batch;
-                        size_t uncompressed_batch_size;
+                    case message:
+                        auto msg = event.message;
+                        switch (msg.type)
+                        {
+                            case MessageType_v2.Records:
+                                Const!(void)[] received_record_batch;
+                                size_t uncompressed_batch_size;
 
-                        this.outer.conn.message_parser.parseBody(
-                            msg.payload, uncompressed_batch_size,
-                            received_record_batch
-                        );
+                                this.outer.conn.message_parser.parseBody(
+                                    msg.payload, uncompressed_batch_size,
+                                    received_record_batch
+                                );
 
-                        (*uncompressed_batch).length = uncompressed_batch_size;
-                        enableStomping(*uncompressed_batch);
+                                (*uncompressed_batch).length = uncompressed_batch_size;
+                                enableStomping(*uncompressed_batch);
 
-                        this.lzo.uncompress(received_record_batch,
-                               *uncompressed_batch);
-                        this.record_stream.addRecords(*uncompressed_batch);
-                        break;
+                                this.lzo.uncompress(received_record_batch,
+                                       *uncompressed_batch);
+                                this.record_stream.addRecords(*uncompressed_batch);
+                                break;
 
-                    case MessageType_v2.Stopped:
-                        this.record_stream.stop();
-                        return;
+                            case MessageType_v2.Stopped:
+                                this.record_stream.stop();
+                                return;
+
+                            case MessageType_v2.Finished:
+                                finished = true;
+                                break;
 
-                    case MessageType_v2.Finished:
-                        finished = true;
+                            default:
+                                assert(false);
+                        }
+                        break;
+                    case signal:
+                        verify(event.signal.code == this.outer.TimeoutSignal);
+                        this.record_stream.stop();
```

Comment:
I don't get what's happening here. The timeout signal may be received. Next, we get to `while(!finished)` and go back to the top of the loop. Or am I missing something?

Reply:
Yes, I was missing `finished = true`.